src/include/codegen/barrier.h

```diff
@@ -0,0 +1,79 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// barrier.h
+//
+// Identification: src/include/codegen/barrier.h
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <atomic>
+#include <boost/thread/barrier.hpp>
+
+#include "common/logger.h"
+#include "codegen/codegen.h"
+#include "codegen/utils/oa_hash_table.h"
+
+namespace peloton {
+namespace codegen {
+
+class Barrier
+{
+
+public:
+    static void InitInstance(Barrier *ins, uint64_t n_workers);
+
+    void MasterWait();
+
+    void Destroy();
+
+    inline bool BarrierWait()
+    {
+        return bar_->wait();
+    }
+
+    inline void WorkerFinish()
+    {
+        --n_workers_;
+    }
+
+    void SetBarrier(boost::barrier *bar) {
+      bar_ = bar;
+    }
+
+    inline void SetWorkerCount(uint64_t n_workers)
+    {
+        n_workers_ = n_workers;
+    }
+
+    inline void MergeToGlobalHashTable(utils::OAHashTable *global_ht, utils::OAHashTable *local_ht)
+    {
+        bool expect = false;
+        while (!global_hash_table_merge_lock_.compare_exchange_strong(expect, true))
+        {
+          expect = false;
+        }
+        global_ht->Merge(local_ht);
+        global_hash_table_merge_lock_.store(false);
+    }
```

Comment:
global_hash_table_merge_lock_ works as a spinning lock here and can be inefficient if contention is high. Can it be implemented by calculating the offset first and merge at the same time like we learned from parallel join algorithm lecture?